src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { cloneDeep } from 'lodash';
+
+export interface Permissions {
+  write?: Principals;
+  read?: Principals;
+  management?: Principals;
+  library_read?: Principals;
+  library_write?: Principals;
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export const PERMISSION_TYPE_READ = 'read';
+export const PERMISSION_TYPE_WRITE = 'write';
+export const PERMISSION_TYPE_MANAGEMENT = 'management';
+export const PERMISSION_TYPE_LIBRARY_READ = 'library_read';
+export const PERMISSION_TYPE_LIBRARY_WRITE = 'library_write';
+
+export const PERMISSION_TYPES: Set<string> = new Set([
```

Comment:
Nit: seems no need to export `PERMISSION_TYPES` as it's not used in other place.